src/coreclr/jit/importer.cpp

```diff
@@ -5486,7 +5486,8 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
         else if (isCastClass)
         {
             // Jit can only inline expand CHKCASTCLASS and CHKCASTARRAY helpers.
-            canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS) || (helper == CORINFO_HELP_CHKCASTARRAY);
```

Comment:
This inlining can be profitable only when the type we are casting to is a class. It is not profitable when the type we are casting to is not a class. For example, it is not profitable when we are casting to `IEnumerable<Canon>` - it is a pure overhead in that case.

Reply:
Presumably the overhead is small or can be eliminated completely if we introduce a slightly stripped version of fallback?

I'll check the stats which path is more often taken in a large service like you requested

Reply:
My primary point was that the inlining of the exact check is not profitable when you can statically tell at JIT time that the target is not a class.

You are right that we can go even further and introduce specialized helper that omits the exact class check. I am not sure whether it is worth it given the overall cost of the helper.

Reply:
For example, we should not be inlining the exact check in:
```
[MethodImpl(MethodImplOptions.NoInlining)]
public IEnumrable<T> GenericCast<T>(object o) => (IEnumerable<T>)o;
```